scripts/upload_s3.py

```diff
@@ -0,0 +1,104 @@
+#!/usr/bin/env python
+
+import argparse
+import json
+import re
+import os
+import boto3
+from json import JSONEncoder
+from pathlib import Path
+from dataclasses import dataclass
+from typing import Dict, Optional, Any
+
+@dataclass
+class TorchBenchData:
+    torch_date: str
+    relpath: str
+
+class TBEncoder(JSONEncoder):
+    def default(self, o):
+        return o.__dict__
+
+S3_BUCKET = "oss-metrics"
+COMMIT_HISTORY: Dict[str, TorchBenchData] = dict()
+
+def get_S3_object_from_bucket(bucket_name: str, object: str) -> Any:
+    s3 = boto3.resource('s3')
+    return s3.Object(bucket_name, object)
+
+def analyze_json_result(json_file: str, relpath: str) -> (str, Optional[TorchBenchData]):
+    with open(json_file, "r") as jfile:
+        data = json.load(jfile)
+    pytorch_version = data["machine_info"]["pytorch_version"]
+    pytorch_date_result = re.search("dev([0-9]{8})", pytorch_version)
+    if not pytorch_date_result:
+        print(f"Found invalid pytorch nightly version {pytorch_version} in {json_file}, skip!")
+        return ("", None)
+    else:
+        pytorch_date = pytorch_date_result.groups()[0]
+        data = TorchBenchData(
+            torch_date = pytorch_date,
+            relpath = relpath)
+        return (pytorch_date, data)
+
+def gen_commit_history(result_dir: str):
+    dirs = map(lambda x: os.path.join(result_dir, x), sorted(os.listdir(result_dir)))
+    dirs = filter(lambda x: os.path.isdir(x), dirs)
+    results = dict()
+    for d in dirs:
+        json_files = map(lambda x: os.path.join(d, x), sorted(os.listdir(d)))
+        # Don't analyze empty files
+        filtered_json_files = filter(lambda x: os.stat(x).st_size, json_files)
+        for json_file in filtered_json_files:
+            relpath = os.path.relpath(json_file, result_dir)
+            (torch_date, tb_data) = analyze_json_result(json_file, relpath)
+            if torch_date:
+                # Only store the latest data
+                results[torch_date] = tb_data
+    out = []
+    for key in sorted(results.keys()):
+        test = dict()
+        test["id"] = key
+        test["result"] = results[key]
+        out.append(test)
+    return out
+
+def upload_s3_file(key, body_file):
+    with open(body_file, "r") as body:
+        print(f"Uploading file {body_file} to S3 key: {key}")
+        obj = get_S3_object_from_bucket(S3_BUCKET, key)
+        obj.put(Body=body.read())
+
+def should_upload(json_path):
+    # Upload non-empty json files
+    return (json_path.endswith(".json") and os.path.exists(json_path) and os.stat(json_path).st_size)
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser(description=__doc__)
```

Comment:
probably good to add some details in __doc__ at the top, explaining what the hierarchy of index and other files is supposed to be and other things about when to use this script.